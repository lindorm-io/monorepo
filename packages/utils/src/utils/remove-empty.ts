import { isArray, isEmpty, isObject } from "@lindorm/is";
import { Dict } from "@lindorm/types";
import { _removeFromArray, _removeFromObject } from "./private";

export function removeEmpty<T extends Array<any> = Array<any>>(array: T): T;
export function removeEmpty<T extends Dict = Dict>(dict: T): T;
export function removeEmpty<T extends Array<any>>(arg: T): T {
  if (isArray(arg)) {
    return _removeFromArray<T>(arg, isEmpty);
  }
  if (isObject(arg)) {
    return _removeFromObject<T>(arg, isEmpty);
  }
  throw new TypeError(`Unsupported type [ ${typeof arg} ]`);
}
