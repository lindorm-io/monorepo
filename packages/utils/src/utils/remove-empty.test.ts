import { removeEmpty } from "./remove-empty";

describe("removeEmpty", () => {
  const testArray = [
    1,
    2,
    3,
    4,
    null,
    undefined,
    "",
    {},
    {
      a: 1,
      b: 2,
      c: null,
      d: undefined,
      e: "",
      f: {},
      g: { a: 1, b: {}, c: null, d: undefined, f: "", h: [] },
    },
    [1, 2, 3, 4, null, undefined, "", {}, { a: 1, b: {} }],
  ];

  const testObject = {
    a: 1,
    b: 2,
    c: null,
    d: undefined,
    e: "",
    f: {},
    g: { a: 1, b: {} },
    h: [
      1,
      2,
      3,
      4,
      null,
      undefined,
      "",
      {},
      { a: 1, b: {} },
      [1, 2, 3, 4, null, undefined, "", {}, { a: 1, b: {}, c: null, d: undefined, f: "", h: [] }],
    ],
  };

  test("should remove empty from array", () => {
    expect(removeEmpty(testArray)).toMatchSnapshot();
  });

  test("should remove empty from object", () => {
    expect(removeEmpty(testObject)).toMatchSnapshot();
  });
});
