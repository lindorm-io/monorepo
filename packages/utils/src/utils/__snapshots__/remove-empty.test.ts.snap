// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeEmpty should remove empty from array 1`] = `
[
  1,
  2,
  3,
  4,
  {},
  {
    "a": 1,
    "b": 2,
    "f": {},
    "g": {
      "a": 1,
      "b": {},
      "h": [],
    },
  },
  [
    1,
    2,
    3,
    4,
    {},
    {
      "a": 1,
      "b": {},
    },
  ],
]
`;

exports[`removeEmpty should remove empty from object 1`] = `
{
  "a": 1,
  "b": 2,
  "f": {},
  "g": {
    "a": 1,
    "b": {},
  },
  "h": [
    1,
    2,
    3,
    4,
    {},
    {
      "a": 1,
      "b": {},
    },
    [
      1,
      2,
      3,
      4,
      {},
      {
        "a": 1,
        "b": {},
        "h": [],
      },
    ],
  ],
}
`;
